Defer response for incoming SIP MESSAGE requests

wosrediinanatour opened the pull request:
For an incoming SIP MESSAGE request, the response is sent immediately. This is a backward-compatible minimal approach for processing incoming SIP MESSAGE transactions asynchronously (see #4589) by deferring the response.

A new account configuration disables immediately response, whereas a new function (PJSUA) can be used for send a response.

It is now possible to respond with a status code != 200.

Even with PJSUA2, deferring a response is possible:
```cpp
void onInstantMessage(pj::OnInstantMessageParam& parameter) override
     {
     pjsip_rx_data* received_data_clone = nullptr;
     // Life-time of msg_info ends at the end of this scope. Copies only msg_info deeply.
     pjsip_rx_data_clone(static_cast<pjsip_rx_data*>(parameter.rdata.pjRxData), 0,
                         &received_data_clone);
     // Life-time ends when PJSIP transaction ends, i.e. after calling pjsua_im_respond_messag(...).
     auto* tsx{pjsip_rdata_get_tsx(static_cast<pjsip_rx_data*>(parameter.rdata.pjRxData))};
     
    //...
     }

// Most probably called from the main thread.
void sendResponse(pjsip_rx_data *received_data_clone, pjsip_transaction *tsx)
    {
    auto text{to_pj_string("OK")};
    pjsua_im_respond_message(&received_data_clone, tsx, 200, &text, nullptr);
    // Life-time of tsx ends very soon.
    }
```

wosrediinanatour commented:
I added a test, where you can see how it is working (both PJSUA and PJSUA2).

wosrediinanatour commented on pjsip/include/pjsua2/account.hpp:
```diff
@@ -1931,6 +1940,127 @@ class FindBuddyMatch
     virtual ~FindBuddyMatch() {}
 };
 
+/**
```
`DeferredResponse` is the token, although not explicit. `DeferredResponse ` is just the ["Rule of 5 (or 6)"](https://www.modernescpp.com/index.php/c-core-guidelines-constructors-assignments-and-desctructors/), so it is nothing special in C++.

The problem is that we need rdata for generating tdata. Or, even if we could make a tdata before (although we don't know the response code), we still would have the same issue:
* rdata has to be cloned, since it becomes destroyed after the fallback
* We need both: non-owning pointer to the transaction and owning pointer to the rdata
* We cannot get rid of the the `DeferredResponse` in the PJSUA2 layer (C++). Because, to be able to save "tokens" to a container, we need to have a move and copy constructors and assignment operators. Because the copy constructors/assignment operators still have to clone "rdata".
* Even if we move the cloning (deep copying) to the lower levels and copy depending on the configuration, we still need it in the user space, when we want to store it. So I would opt against it for locality.

The only improvement, I would come up, is making a similar "class" in C as well (PJUSA):
* A struct that combines pointers `pjsip_rx_data*` and `pjsip_transaction*`.
* Functions that shallow copies and deep copy
* Note that the "rather lengthy and convoluted" code would then be in the user space (as now).
* The struct could be put to the callbacks as you suggested. For C++ this would be OK. But for C (PJSUA) this would mean to add another version of the callback. Something I frankly don't like much because it is not backward compatible.

sauwming replied:
> * pjsua2 will manage the DeferredResponse under the hood (so it does not need to be visible to the user).

What I meant was that DeferredResponse to be moved to `account.cpp`, so all the things you mentioned will still be done, but will be transparent to the user.

sauwming commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -4782,6 +4782,21 @@ typedef struct pjsua_acc_config
+    *
+    * Default: PJ_FALSE (immediate response for backward compatibility)
+    */
+    pj_bool_t        process_sip_message_async;
```
Ah okay, I forgot about `on_call_rx_reinvite()` (even though I was the one that implemented it :) ). I believe the context that time was for async third party media.

`auto_process_sip_message` sounds fine to me (`auto_send_200_for_sip_message` seems too long).

sauwming commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -7283,6 +7298,26 @@ PJ_DECL(pj_status_t) pjsua_im_typing(pjsua_acc_id acc_id,
+ *
+ * @return              PJ_SUCCESS on success, or the appropriate error code.
+ */
+PJ_DECL(pj_status_t) pjsua_im_respond_message(pjsip_rx_data *rdata,
```
If it's not necessary, then it doesn't have to be tied to "instant messaging".

For example, since we already have `pjsua_acc_send_request()`, adding `pjsua_acc_send_response()` seems to make sense. Or you have other ideas?

wosrediinanatour commented:
@sauwming: Applied the changes - PTAL https://github.com/pjsip/pjproject/compare/4ec864123be3a04598b5bc7242ae223690371eff..0e6bbf65902801f0e5dbffacf7e9203cc7842fc6

sauwming reviewed (approved):
Just a few minors. That's all from me.

wosrediinanatour commented:
I split unit tests into header and implementation. Since I had to change several lines, I also added a second unit test for testing the "immediate response". PTAL at the changes https://github.com/pjsip/pjproject/compare/ccd3498652137c8592083d9453895cb5d85c7dd5..0a0d875c27a310a378b70c143e36e8484bc95bec

wosrediinanatour commented:
Additional minor changes pushed: https://github.com/pjsip/pjproject/compare/0a0d875c27a310a378b70c143e36e8484bc95bec..af81002ad4ca2c1b871344d33377c7861e9a7d6c

nanangizz commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -5116,6 +5131,33 @@ PJ_DECL(pj_status_t) pjsua_acc_get_config(pjsua_acc_id acc_id,
+                                             int st_code,
+                                             const pj_str_t *st_text,
+                                             const pjsip_hdr *hdr_list);
+
```
I think it is better to use `const pjsua_msg_data *msg_data` instead of `hdr_list` if possible.

nanangizz commented on pjsip/src/pjsua-lib/pjsua_im.c:
```diff
@@ -304,13 +319,30 @@ static pj_bool_t im_on_rx_request(pjsip_rx_data *rdata)
+    acc_id = pjsua_acc_find_for_incoming(rdata);
+    
+    if (acc_id != PJSUA_INVALID_ID && pjsua_var.acc[acc_id].cfg.auto_process_sip_message == PJ_FALSE
+            && is_typing_indication(rdata) == PJ_FALSE) {
```
Minor:
- Coding style for [brace in multiline statement](https://docs.pjsip.org/en/latest/get-started/coding-style.html#use-k-r-style-brace-placement), the brace should be in a separate line.
- While in coding style, please also recheck for [line length limit](https://docs.pjsip.org/en/latest/get-started/coding-style.html#limit-line-length) in the whole PR code.

nanangizz commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -4782,6 +4782,21 @@ typedef struct pjsua_acc_config
+    *
+    * When set to PJ_FALSE (manual processing), incoming MESSAGE requests
+    * will create UAS transactions and defer response sending. Applications
+    * must use pjsua_acc_send_response() to send responses manually.
```
How about also adding a brief hint about the required steps for the app in the docs, for example:

> app must clone the request data (using ...) and obtain the transaction instance (using ...) from the `on_pager()/on_pager2()` callback, which is then used to send the response using `pjsua_acc_send_response()`.

nanangizz commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -4782,6 +4782,21 @@ typedef struct pjsua_acc_config
+    *
+    * Default: PJ_TRUE (automatic response for backward compatibility)
+    */
+    pj_bool_t        auto_process_sip_message;
```
IMHO `auto_process` sounds somewhat ambigous as the libray is merely sending a response automatically, not fully processing it, so how about `auto_respond` instead?

wosrediinanatour commented:
Incorporated @nanangizz feedback in https://github.com/pjsip/pjproject/compare/3aa2faf28cf0cc6734d0a7494f710ca0905e177f..2a859d0ed260d2a92fcc39c132fa9e857b55ca09.

wosrediinanatour commented:
I hope that https://github.com/pjsip/pjproject/compare/2a859d0ed260d2a92fcc39c132fa9e857b55ca09..3f81713293c421f5cf303b5de760470836934c03 now fixes all ill-formatted code.
